bee-network/bee-autopeering/src/peer/peer_id.rs

```diff
@@ -0,0 +1,218 @@
+#[derive(Copy, Clone)]
+pub struct PeerId {
+    // An ED25519 public key.
+    public_key: PublicKey,
+    // The SHA256 hash of the ED25519 public key.
+    id_bytes: [u8; hash::SHA256_LEN],
+    // The corresponding `libp2p` PeerId.
+    libp2p_peer_id: libp2p_core::PeerId,
+    // The displayed Base58 representation.
+    base58_peer_id: [char; DISPLAY_LENGTH],
+}
+
+impl PeerId {
+    /// Generates a new random `PeerId`.
+    pub fn generate() -> Self {
+        let private_key = PrivateKey::generate().expect("error generating private key");
+
+        Self::from_public_key(private_key.public_key())
+    }
+
+    /// Creates a peer identity from an ED25519 public key.
+    pub fn from_public_key(public_key: PublicKey) -> Self {
+        let id_bytes = hash::sha256(public_key.as_ref());
+        let libp2p_peer_id = to_libp2p_peer_id(public_key);
+
+        let mut base58_peer_id: [char; DISPLAY_LENGTH] = Default::default();
+        base58_peer_id.copy_from_slice(
+            libp2p_peer_id.to_base58()[..DISPLAY_LENGTH]
+                .chars()
+                .collect::<Vec<_>>()
+                .as_slice(),
+        );
+
+        Self {
+            id_bytes,
+            public_key,
+            libp2p_peer_id,
+            base58_peer_id,
+        }
+    }
+
+    /// Returns the public key associated with this identity.
+    pub fn public_key(&self) -> &PublicKey {
+        &self.public_key
+    }
+
+    /// Returns the actual bytes representing this id.
+    pub fn id_bytes(&self) -> &[u8; hash::SHA256_LEN] {
+        &self.id_bytes
+    }
+
+    /// Turns this `PeerId` into a `libp2p_core::PeerId`.
+    pub fn into_libp2p_peer_id(self) -> libp2p_core::PeerId {
+        self.libp2p_peer_id
+    }
+}
+
+/// Creates the corresponding `libp2p_core::PeerId` from a crypto.rs ED25519 public key.
+pub fn to_libp2p_peer_id(public_key: PublicKey) -> libp2p_core::PeerId {
+    libp2p_core::PeerId::from_public_key(to_libp2p_public_key(public_key))
+}
+
+/// Creates the corresponding `libp2p_core::PublicKey` from a crypto.rs ED25519 public key.
+pub fn to_libp2p_public_key(public_key: PublicKey) -> libp2p_core::PublicKey {
+    libp2p_core::PublicKey::Ed25519(
+        libp2p_core::identity::ed25519::PublicKey::decode(public_key.as_ref())
+            .expect("error decoding ed25519 public key from bytes"),
+    )
+}
+
+impl Eq for PeerId {}
+impl PartialEq for PeerId {
+    fn eq(&self, other: &Self) -> bool {
+        self.id_bytes == other.id_bytes
+    }
+}
+impl Hash for PeerId {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.id_bytes.hash(state);
+    }
+}
+
+impl fmt::Debug for PeerId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let s = &bs58::encode(&self.id_bytes).into_string();
+
+        f.debug_struct("PeerId")
+            .field("public_key", &bs58::encode(self.public_key).into_string())
+            .field("id", &s)
+            .finish()
```

Comment:
I'm slightly confused, don't we have the bs58 computed already ?

Reply:
I fully understand your confusion, but those are 2 different things actually:
(a) ED25519 public_key -> libp2p_peer_id -> base58 -> base58[..DISPLAY_LENGTH]
(b) ED25519 public_key -> base58

Reply:
So here I want to print the actual public key everything is derived of.